### ISSUE TITLE
Please provide a "pb_seek" interface

### ISSUE THREAD

**Issue by GoogleCodeExporter** [nanopb/nanopb#137]
```
Describe the requested feature:

A "pb_seek" interface that tells nanopb to update its internal state to match 
some other buffer write done outside of the nanopb framework.

In what situation would the feature be useful:

Some functions already write into memory; writing into a temporary buffer and 
then calling pb_write from that buffer into the stream is wasteful. The 
alternative of writing directly into the stream and than calling pb_write 
"in-place" also works, but is also wasteful (reading/writing the stream memory 
with no modification).

Example functions:

+ memset/memcpy/strcpy
+ printf and friends
+ cryptographic and other transformative functions that can read from one piece 
of memory and write to another piece of memory

In all above cases, it would be nice to have a "pb_seek" function that does the 
same as "pb_write" (for updating stream->state, stream->bytes_written, etc.) 
except for the memcpy part of it (since it was already done "out-of-band").

I've attached an example implementation. An actual integrated patch could 
perhaps be slightly more efficient by having buf_write call buf_seek, then 
copying the memory.

I think such a feature only makes sense for PB_BUFFER_ONLY since the meaning of 
"stream->state" is very specific to the buffer implementation and could mean 
something completely different for a custom callback.
```

Original issue reported on code.google.com by `robt...@google.com` on 12 Dec 2014 at 11:23

Attachments:
- [nanopb-seek.patch.txt](https://storage.googleapis.com/google-code-attachments/nanopb/issue-137/comment-0/nanopb-seek.patch.txt)

**Comment by jkmnt** [comment]
I'd like to share on the issue.

The cost of two-pass serialization is high in case of many nested submessages. Even worse, in out case data may change between passes.  Our streams are supporting seeks.

Nanopb currently isn't supporting seeks, so we implemented the hybrid solution. 

Nanopb is serializing the leaf objects without the nested submessages and doing it fine. 
Top messages are serialized manually with the seeks supported and no perfomance penalty.

Our API for the submessage serialization is like:

```c
bool pblite_encode_bytes_begin(ostream_t *substream, ostream_t *base);
bool pblite_encode_bytes_end(ostream_t *substream, ostream_t *base);
```

We open the bytes substream (spawned from the base stream), write into the substream and close it.

'begin' function seeks a few bytes (2 by default) ahead on a base stream, reserving it for the length field and initializes the substream.

'end' function tells() on a substream obtaining the size, seeks the base back to a length field, encodes length and restores base stream position.

The length is encoded in a subnormal form, i.e. the varint is padded with 0x80 until the fixed field size reached. It's perfectly legal and one byte memory overhead is neglegible in out case. 2 bytes length is good for up to 16k submessage size.

The proposed enhancement for nanopb would be the 'seek' callback on output streams. If seek is present, use it for serialization of submessages, otherwise fallback to the two-pass serializarion.

**Comment by PetteriAimonen** [comment]
Yeah, I've been considering the same to avoid double-serialization.

As an alternative for a raw seek interface, I've considered exposing some kind of "encode length-prefixed subitem" interface. But I'm not sure which is better. On other hand, "encode length-prefixed subitem" could be used with streams also with some kind of temporary buffer if the submessage size is small enough.

Padding with 0x80 is indeed technically legal, but at least for plain memory buffers (`pb_ostream_from_buffer()`), I think I'd like a `memmove()` to make the output canonical. Otherwise testing is made more complicated by the small differences in binary data. I think it could predict the message size based on what is known in generation time, so most of the time the extra move could be skipped, and if it is needed, the overhead is still less than double-encoding.

**Comment by tadiparthi** [comment]
a simple memcpy based one pass encoding approach.
instead of writing to a null stream, it will write to a temporary stream and then copies the encoded data to the original stream

[nanopb_onepass_encoding.txt](https://github.com/nanopb/nanopb/files/4811745/nanopb_onepass_encoding.txt)
